easybuild/easyblocks/m/mesa.py

```diff
@@ -0,0 +1,61 @@
+##
+# Copyright 2009-2020 Ghent University
+#
+# This file is part of EasyBuild,
+# originally created by the HPC team of Ghent University (http://ugent.be/hpc/en),
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://www.vscentrum.be),
+# Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# https://github.com/easybuilders/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for installing Mesa, implemented as an easyblock
+
+@author: Andrew Edmondson (University of Birmingham)
+"""
+
+from easybuild.easyblocks.generic.mesonninja import MesonNinja
+from easybuild.tools.systemtools import POWER, X86_64, get_cpu_architecture, get_cpu_features
+
+
+class EB_Mesa(MesonNinja):
+    def configure_step(self, cmd_prefix=''):
+        """
+        Customise the configopts based on the platform
+        """
+        arch = get_cpu_architecture()
+        if arch == X86_64:
+            self.cfg.update('configopts', "-Dgallium-drivers='swrast,swr'")
+        elif arch == POWER:
+            self.cfg.update('configopts', "-Dgallium-drivers='swrast'")
+
+        features = set(get_cpu_features())
+        arches = []
+        if 'avx' in features or 'avx1.0' in features:
+            arches.append('avx')
+        if 'avx2' in features:
+            arches.append('avx2')
+        if 'avx512f' in features:
+            # AVX-512 Foundation - introduced in Skylake
+            arches.append('skx')
+        if 'avx512er' in features:
+            # AVX-512 Exponential and Reciprocal Instructions implemented in Knights Landing
+            arches.append('knl')
+        if arches:
+            self.cfg.update('configopts', "-Dswr-arches=%s" % ','.join(arches))
```

Comment:
Shjouldn't this whole block be inside the "if arch == X86_64"

Reply:
Judging from the name I'd expect this only to be used when `swr` is built, so I guess yes.

> Is cross-compilation "supported" by EasyBuild? Not using it, I wouldn't know. So many things auto-detect their CPU features that I'd have thought it would mostly not build with the right support - but I've not tried it.

Kinda. For example we do have 3 installation trees: Power, sandy bridge, haswell. The haswell is also used for (a few) newer ones. So while this is no real cross compilation it is compiling for support on the maximal common arch. So I'd include the arches unconditionally based on 

>  they are selected at runtime

Reply:
lol - this is what it was like earlier before https://github.com/easybuilders/easybuild-easyblocks/pull/1892#pullrequestreview-371226403

What's the conclusion @Flamefire and @akesandgren ?

Reply:
I'd add them always if swr is built or do that in the EC. I have the feeling a compiler might choke if e.g. knl is added but the compiler is to old. Not sure how smart the configure is. If it is smart enough: Add always in EasyBlock, else in EasyConfig

Reply:
Not sure what you mean by that. That whole block of checking for avx,skx,knl are only valid if it is a X86_64 architecture in the first place, so why do it even if the base arch is POWER?

Reply:
Sorry for the confusion:
- `swr-arches` (most likely) applies only when building swr, hence yes this block need to be moved inside the X86_64 check
- I would not do the arch check at all and simply add all available archs to `swr-arches` inside X86_64 to allow selection at runtime
- If the configure of Mesa would add flags that are incompatible with older compilers when e.g. `swr-arches=knl` is set (happens of Mesa doesn't check those flags before adding, but I don't know if it does check) then adding `swr-arches` should be done in an EasyConfig instead (where compiler support for such flags is known)

Reply:
Ok I took the time to verify the last point: Mesa *does* add the flags unconditionally. Hence if `swr-arches=knl` is given and the compiler doesn't support this flag the configure will fail.

-> Add those archs via EasyConfig

Reply:
well compiler support is sort of knowable in the easyblock too by checking compiler version

Reply:
compiler support for `swr-arches` is not an issue right now because this easyblock for `mesa` will only be used with recent toolchains (i.e. old easyconfigs for Mesa will stay with `configuremake`). In the future, if there are compiler dependent `configopts`, we can just add specific conditionals to the `mesa` easyblock.

Reply:
OK - implemented in the suggested changes below from @smoors :+1: